### PR TITLE
StringSegment AsSpan and AsMemory

### PR DESCRIPTION
**Pull request by gfoidl**
Fixes https://github.com/aspnet/Home/issues/2890

### CODE REVIEW

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 1]
```diff
@@ -77,6 +78,25 @@ private static Exception GetInvalidArgumentException(string buffer, int offset,
             return ThrowHelper.GetArgumentException(ExceptionResource.Argument_InvalidOffsetLength);
         }
 
+        /// <summary>
+        /// Initializes an instance of the <see cref="StringSegment"/> struct.
+        /// </summary>
+        /// <param name="span">The original <see cref="Span{T}"/> used as buffer.</param>
+        public unsafe StringSegment(Span<char> span)
+        {
+            Buffer = new string((char*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(span)));
```
Comment (benaadams) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: The use of `AsPointer` risks a GC hole as the spans data may not be pinned; you'd have to fix it instead to get the pointer. Also should use an override that takes length (or will have issues with null termination)

```csharp
fixed (char* src = &MemoryMarshal.GetReference(span))
{
    Buffer = new string(src, span.Length);
}
```

Comment (gfoidl) [reply src/Microsoft.Extensions.Primitives/StringSegment.cs]: Thanks for the hint!

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 2]
```diff
@@ -77,6 +78,29 @@ private static Exception GetInvalidArgumentException(string buffer, int offset,
             return ThrowHelper.GetArgumentException(ExceptionResource.Argument_InvalidOffsetLength);
         }
 
+        /// <summary>
+        /// Initializes an instance of the <see cref="StringSegment"/> struct.
+        /// </summary>
+        /// <param name="span">The original <see cref="Span{T}"/> used as buffer.</param>
+        public unsafe StringSegment(Span<char> span)
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: I'm not sure this direction is very useful as it allocates which isn't really the point here. Going from StringSegment -> Span\<char\> or ReadOnlyMemory<char> makes more sense than this direction. I'd get rid of this ctor.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 3]
```diff
@@ -77,6 +78,29 @@ private static Exception GetInvalidArgumentException(string buffer, int offset,
...
+
+            Offset = 0;
+            Length = span.Length;
+        }
+
+        /// <summary>
+        /// Initializes an instance of the <see cref="StringSegment"/> struct.
+        /// </summary>
+        /// <param name="memory">The original <see cref="Memory{T}"/> used as buffer.</param>
+        public StringSegment(Memory<char> memory) : this(memory.Span)
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Same here, I'm not sure this is useful.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 2 above)
Comment (gfoidl) [reply src/Microsoft.Extensions.Primitives/StringSegment.cs]: I reverted the commit that adds the ctors, updated the PR.

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 4]
```diff
@@ -136,6 +136,18 @@ public bool HasValue
             }
         }
 
+        /// <summary>
+        /// Gets a <see cref="ReadOnlySpan{T}"/> from the current <see cref="StringSegment"/>.
+        /// </summary>
+        /// <returns>The <see cref="ReadOnlySpan{T}"/> from this <see cref="StringSegment"/>.</returns>
+        public ReadOnlySpan<char> AsSpan() => Buffer.AsReadOnlySpan(Offset, Length);
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Should this be AsReadOnlySpan?

File: src/Microsoft.Extensions.Primitives/StringSegment.cs [hunk 5]
```diff
@@ -136,6 +136,18 @@ public bool HasValue
...
+
+        /// <summary>
+        /// Gets a <see cref="ReadOnlyMemory{T}"/> from the current <see cref="StringSegment"/>.
+        /// </summary>
+        /// <returns>The <see cref="ReadOnlyMemory{T}"/> from this <see cref="StringSegment"/>.</returns>
+        public ReadOnlyMemory<char> AsMemory() => Buffer.AsReadOnlyMemory(Offset, Length);
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Primitives/StringSegment.cs]: Same here, should this be AsReadOnlyMemory?

**Comment by davidfowl** [comment]
I'm not sure what the methods should be called https://github.com/dotnet/corefx/blob/502ad71e89b0aa3a51d07e99204befb3c85573e8/src/System.Memory/ref/System.Memory.cs#L14

@terrajobst @KrzysztofCwalina 

What did we land on?

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply src/Microsoft.Extensions.Primitives/StringSegment.cs]: Remember, we are renaming these APIs on String to AsSpan and AsMemory, i.e. read-only being implied.

Comment (davidfowl) [reply src/Microsoft.Extensions.Primitives/StringSegment.cs]: Ok so this PR is good to go then.

**Comment by davidfowl** [comment]
Thanks!

File: src/Microsoft.Extensions.Primitives/StringSegment.cs (hunk 4 above)
Comment (gfoidl) [reply src/Microsoft.Extensions.Primitives/StringSegment.cs]: I think https://github.com/dotnet/corefx/pull/27328 is the mentioned rename. Version `4.5.0-preview2-26227-01` of `System.Memory` already has these renamed methods. Shall I submit an updated PR to be up to date with this -- or create an issue to track the renaming here?

**Pull request merged**